src/plugins/mer/meremulatordevice.cpp

```diff
@@ -462,31 +527,54 @@ void MerEmulatorDevice::updateConnection() const
 
 void MerEmulatorDevice::updateAvailableDeviceModels()
 {
+    m_availableDeviceModels.clear();
+    m_sdkDeviceModelNames.clear();
+
+    // Reading predefined emulators
+
     //! \todo Does not support multiple (different) emulators (not supported at other places anyway).
     PersistentSettingsReader reader;
-    if (!reader.load(globalSettingsFileName()))
-        return;
+    if (reader.load(globalSettingsFileName())) {
+        const QVariantMap data = reader.restoreValues();
 
-    QVariantMap data = reader.restoreValues();
+        int version = data.value(QLatin1String(MER_DEVICE_MODELS_FILE_VERSION_KEY), 0).toInt();
+        if (version < 1) {
+            qWarning() << "Invalid configuration version: " << version;
+            return;
+        }
 
-    int version = data.value(QLatin1String(MER_DEVICE_MODELS_FILE_VERSION_KEY), 0).toInt();
-    if (version < 1) {
-        qWarning() << "Invalid configuration version: " << version;
-        return;
-    }
+        int count = data.value(QLatin1String(MER_DEVICE_MODELS_COUNT_KEY), 0).toInt();
+        for (int i = 0; i < count; ++i) {
+            const QString key = QString::fromLatin1(MER_DEVICE_MODELS_DATA_KEY) + QString::number(i);
+            if (!data.contains(key))
+                break;
 
-    int count = data.value(QLatin1String(MER_DEVICE_MODELS_COUNT_KEY), 0).toInt();
-    for (int i = 0; i < count; ++i) {
-        const QString key = QString::fromLatin1(MER_DEVICE_MODELS_DATA_KEY) + QString::number(i);
-        if (!data.contains(key))
-            break;
+            const QVariantMap deviceModelData = data.value(key).toMap();
+            MerEmulatorDeviceModel deviceModel;
+            deviceModel.fromMap(deviceModelData);
+
+            m_availableDeviceModels.insert(deviceModel.name(), deviceModel);
+            m_sdkDeviceModelNames.insert(deviceModel.name());
+        }
+    }
 
-        const QVariantMap deviceModelData = data.value(key).toMap();
-        MerEmulatorDeviceModel deviceModel;
-        deviceModel.fromMap(deviceModelData);
+    // Reading user defined emulators
+    QSettings *settings = ICore::settings();
+    settings->beginGroup(QLatin1String(MER_EMULATOR_SETTINGS_CATEGORY));
```

Comment:
Save and read should be implemented in one class, so this code should be moved to MerSettings unless I missed something.

Reply:
It was there before you said: [https://github.com/sailfishos/sailfish-qtcreator/pull/141#discussion_r237068049](https://github.com/sailfishos/sailfish-qtcreator/pull/141#discussion_r237068049). So all related functionality moved. I have to introduce new function like `static QHash<QString, MerEmulatorDeviceModel> userDefinedDeviceModels()` in mersettings.cpp, if I should read settings there.

Reply:
This code moved to MerSettings